zng/builder.go

```diff
@@ -48,3 +54,168 @@ func (b *Builder) Build(zvs ...zcode.Bytes) *Record {
+	// are not sorted, but we don't do this yet.
+	if typErr := b.rec.TypeCheck(); typErr != nil {
+		// type error overrides ErrIncomplete
+		err = typErr
+	}
+	return &b.rec, err
+}
+
+func (b *Builder) parseRecord(typ *TypeRecord, in []string) ([]string, error) {
+	var err error
+	for _, col := range typ.Columns {
+		if len(in) == 0 {
+			err = ErrIncomplete
+			b.appendUnset(col.Type)
+			continue
+		}
+
+		switch v := AliasedType(col.Type).(type) {
+		case *TypeRecord:
+			b.BeginContainer()
+			in, err = b.parseRecord(v, in)
+			if err != nil && err != ErrIncomplete {
+				return nil, err
+			}
+			b.EndContainer()
+		case *TypeArray:
+			b.BeginContainer()
+			if err = b.parseArray(v, in[0]); err != nil {
+				return nil, err
+			}
+			b.EndContainer()
+			in = in[1:]
+		case *TypeSet:
+			b.BeginContainer()
+			if err = b.parseSet(v, in[0]); err != nil {
+				return nil, err
+			}
+			b.EndContainer()
+			in = in[1:]
+		case *TypeUnion:
+			// XXX need a value syntax that indicates which type to use
+			return nil, errors.New("union values not yet supported")
+		default:
+			if err = b.parsePrimitive(v, in[0]); err != nil {
+				return nil, err
+			}
+			in = in[1:]
+		}
+	}
+	return in, err
+}
+
+func (b *Builder) parseArray(typ *TypeArray, in string) error {
+	inner := InnerType(AliasedType(typ))
+	if len(in) == 0 {
+		return nil
+	}
+	if in[0] != '[' || in[len(in)-1] != ']' {
+		return ErrBadFormat
+	}
+	in = in[1 : len(in)-1]
+	if len(in) == 0 {
+		// empty array
+		b.appendUnset(inner)
+		return nil
+	}
+	//XXX for now just use simple comma rule, which means comman
+	// cannot be embedded in a set value here.  we need a recursive
+	// descent parser like the type parser to d this correctly
+	for _, val := range strings.Split(in, ",") {
+		switch v := inner.(type) {
+		case *TypeRecord:
+			// XXX need a syntax for this, and recursive descent
+			return errors.New("array of records value not yet supported")
+		case *TypeArray:
+			// XXX need a syntax for this, and recursive descent
+			return errors.New("array of arrays value not yet supported")
+		case *TypeSet:
+			// XXX need a syntax for this, and recursive descent
+			return errors.New("array of sets value not yet supported")
+		case *TypeUnion:
+			// XXX need a syntax for this, and recursive descent
+			return errors.New("array of unions value not yet supported")
+		default:
+			b.parsePrimitive(v, val)
+		}
+	}
+	return nil
+}
+
+func (b *Builder) parseSet(typ *TypeSet, in string) error {
+	inner := InnerType(AliasedType(typ))
+	if len(in) == 0 {
+		return nil
+	}
+	if in[0] != '[' || in[len(in)-1] != ']' {
+		return ErrBadFormat
+	}
+	in = in[1 : len(in)-1]
+	if len(in) == 0 {
+		// empty set
+		b.appendUnset(inner)
+		return nil
+	}
+	if IsContainerType(inner) {
+		return &RecordTypeError{Name: "<set>", Type: typ.String(), Err: ErrNotPrimitive}
+	}
+	//XXX for now just use simple comma rule, which means comman
```

Comment:
I have a question: A comma is different than the `;` separator we use to separate items in tzng, why not use a semicolon?

Reply:
I don't think tzng syntax is the right thing.  The question is what is the zql syntax for literals of the container types?